# Mandatory Assignment 3

#### MAT-2201 Numerical Methods

 Daniel Elisabethsønn Antonsen, UiT

Importing libraries and modules needed for the assignment.

In [1]:
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
plt.style.use('seaborn-whitegrid')

## Problem 1

*Solve Problem 3 from the MAT-2201 exam from December 2019.*

A quantity $y(t)$ depends on the time variable $t$, is measured in four time points. The measurements are given in the table below.

| $t_i$ for $i = 1,...,4$ | $y_i$ for $i = 1,...,4$ |
|------------------------|-------------------------|
| 0                    | 1                     |
| 1                    | 2                     |
| 2                    | 7                     |
| 3                    | 5                     |

#### (a)

We are here asked to find a polynomial $P(t)$ of degree $\leq 2$ that goes through the **first three** points
$$
(t_1, y_1), (t_2, y_2), (t_3, y_3).
$$

To find the polynomial $P(t)$ we can make use of Lagrange interpolation, which is defined of a degree $leq 2$ as
$$
P_2 (t) = y_1\frac{(t - t_2) (t - t_3)}{(t_1 - t_2)(t_1 - t_3)} + y_2\frac{(t - t_1) (t - t_3)}{(t_2 - t_1)(t_2 - t_3)} + y_3\frac{(t - t_1) (t - t_2)}{(t_3 - t_1)(t_3 - t_2)}
$$
And so by using the values given in the table we can compute the polynomial as

\begin{align*}
P_2 (t) &= 1 \frac{(t - 1) (t - 2)}{(0 - 1)(0 - 2)} + 2 \frac{(t - 0) (t - 2)}{(1 - 0)(1 - 2)} + 7 \frac{(t - 0) (t - 1)}{(2 - 0)(2 - 1)} \\
&= \frac{1}{2} t^2 - \frac32 t + 1 - 2t^2 + 2t + \frac{7}{2} t^2 - \frac{7}{2} t \\
&= \underline{2t^2 - 3t + 1}
\end{align*}

And so a polynomial $P(t)$ of degree $\leq 2$ is given by
$$
\boxed{P(t) = 2t^2 - 3t + 1}
$$



#### (b)

We are here asked to find a polynomial $S(t)$ of degree $\leq 2$ that fits the **four** points 
$$
(t_1, y_1), (t_2, y_2), (t_3, y_3), (t_4, y_4).
$$
i.e. that **minimizes** the sum given by
$$
\sum_{i=1}^4 (S(t_i) - y_i)^2.
$$

The sum given in the problem, is the squared error or residuals. We can minimize this by using the method of least squares, and if we are given a data set of $m$ data points, can be implemented in three steps.

(1) Choose a model, a parametrized model. This model in our case, will be of the form
$$
S(t) = y = c_1 + c_2 t + c_3 t^2
$$
(2) We then force the model to fit the given data set by substituting in the data points. Each of the data points will then form an equation where all the unknowns are the parameters $c_1$, $c_2$ and $c_3$. This will then form the system of equation given in matrix form as 
$$
A C = b
$$
Where $t$ represents the unknown parameters.  
(3) As the last step, we solve the normal equations which gives us the least squares estimate for the parameters $\bar{C}$. The normal equation in found by computing 
$$
A^T A C = A^T b
$$
Or rewriting to 
$$
C = (A^T A)^{-1} A^T b
$$
where $C = \begin{pmatrix} c_1 & c_2 & c_3 \end{pmatrix}^T$. And so, we can start by considering step $1$ and choose a model of the form 
$$
S(t) = y = c_1 + c_2 t + c_3 t^2
$$
Which leads to the system of equations given as

\begin{align*}
y_1 &= c_1 + 0 \cdot c_2 + 0 \cdot c_3 = 1 \\
y_2 &= c_1 + c_2 + c_3 = 2 \\
y_3 &= c_1 + 2 c_2 + 4 c_3 = 7 \\
y_4 &= c_1 + 3 c_2 + 9 c_3 = 5
\end{align*}

Which in matrix form can be written as 
$$
\begin{pmatrix} 1 & 0 & 0 \\ 1 & 1 & 1 \\ 1 & 2 & 4 \\ 1 & 3 & 9 \end{pmatrix}
\begin{pmatrix} c_1 \\ c_2 \\ c_3 \end{pmatrix} 
= 
\begin{pmatrix} 1 \\ 2 \\ 7 \\ 5 \end{pmatrix}
$$

We can here begin to solve and so solve the equations by finding the matricies $A^T A$ and $A^T $, before solving the normal equation by hand. Or we can simply use python code to solve the equations by implementing the matricies found above. We will here use the second choice for solving and impelment python code for solving for the coefficients.

We start by implementing the matricies as arrays

In [2]:
A = np.array([[1, 0, 0], [1, 1, 1], [1, 2, 4], [1, 3, 9]])
b = np.array([[1], [2], [7], [5]])

Now that we have implemented the arrays for the known matricies, we can now solvbe the normal equations given as
$$
C = (A^T A)^{-1} A^T b
$$
And so

In [3]:
C = np.linalg.inv(A.T@A)@A.T@b
print(f"The solution for the normal equation is {C.T}^T")

The solution for the normal equation is [[ 0.45  3.95 -0.75]]^T


And so, with this result we have that the coefficients are given as 
$$
C 
= 
\begin{pmatrix} c_1 \\ c_2 \\ c_3 \end{pmatrix}
=
\begin{pmatrix} 0.45 \\ 3.95 \\ -0.75 \end{pmatrix}
$$

Using this result, we get that a polynomial given by
$$
S(t) = -0.75 t^2 + 3.95 t + 0.45
$$
which minimizes the the sum 
$$
\sum_{i = 1}^4 (S(t_i) - y_i)^2.
$$


#### (c)

We are given the matrix
$$
B = \begin{bmatrix} 1 & 0 \\ 1 & 1 \\ 1 & 2 \\ 1 & 3 \end{bmatrix}
$$
and is asked to find the **reduced** QR factorization of B, i.e. express B as 
$$
B = QR
$$
where 
$$
Q = \begin{bmatrix} \mathbf{q}_1 & \mathbf{q}_2 \end{bmatrix}
$$
is a $4\times 2$ matrix with orthonormal columns ($\mathbf{q}_i \cdot \mathbf{q}_j = 0$ if $i \neq j$ and $1$ if $i = j$), and $R$ is an upper-triangular $2\times 2$ matrix.

We can start by computing the QR factorization we can start by setting 
$$
y_1 = B_1 = \begin{bmatrix} 1 \\ 1 \\ 1 \\ 1 \end{bmatrix}
$$
Then we can further compute the $2$-norm or the Euclidean norm, which gives
$$
r_{1 1} = ||y_1||_2 = \sqrt{1^2 + 1^2 + 1^2 + 1^2} = \sqrt{4} = 2
$$
And so, the first unit vector is then found by
$$
q_1 = \frac{y_1}{||y_1||_2} 
= \frac12 \begin{bmatrix} 1 \\ 1 \\ 1 \\ 1 \end{bmatrix}
= \begin{bmatrix} 1/2 \\ 1/2 \\ 1/2 \\ 1/2 \end{bmatrix}
$$
Now that we have found the first unit vector or the orthonormal vector, we can continue by finding the second unit vector. This can be found by
$$
y_2 = B_2 - q_1 q_1^T B_2 
= \begin{bmatrix} 0 \\ 1 \\ 2 \\ 3 \end{bmatrix} - \begin{bmatrix} 3/2 \\ 3/2 \\ 3/2 \\ 3/2 \end{bmatrix}
= \begin{bmatrix} -3/2 \\ -1/2 \\ 1/2 \\ 3/2 \end{bmatrix}
$$
We can now compute the $2$-norm for $y_2$ which gives us 
$$
r_{2 2} = ||y_2||_2 = \sqrt{(-3/2)^2 + (-1/2)^2 + (1/2)^2 + (3/2)^2} = \sqrt{5}
$$
And so the second unit vector is given by 
$$
q_2 = \frac{y_2}{||y_2||_2} = \frac{1}{\sqrt{5}} \begin{bmatrix} -3/2 \\ -1/2 \\ 1/2 \\ 3/2 \end{bmatrix}
= \begin{bmatrix} -\frac{3}{2\sqrt{5}} \\ -\frac{1}{2\sqrt{5}} \\ \frac{1}{2\sqrt{5}} \\ \frac{3}{2\sqrt{5}} \end{bmatrix}
$$
We can also find the element which is not on the diagonal by, and since $R$ is upper-triangular we only need to consider 
$$
r_{1 2} = q_1^T B_2
$$
And so $r_{1 2} = 3$, which gives that the $QR$ factorization for the matrix $B$ is given in matrix form as
$$
\boxed{B = \begin{bmatrix} 1 & 0 \\ 1 & 1 \\ 1 & 2 \\ 1 & 3 \end{bmatrix}
= \begin{bmatrix} 1/2 & -\frac{3}{2\sqrt{5}} \\ 1/2 & -\frac{1}{2\sqrt{5}} \\ 1/2 & \frac{1}{2\sqrt{5}} \\ 1/2 & \frac{3}{2\sqrt{5}} \end{bmatrix}
\begin{bmatrix} 2 & 3 \\ 0 & \sqrt{5} \end{bmatrix}
= Q R}
$$



#### (d)

We are here asked to find a polynomial $U(t)$ of degree $\leq 1$ that fits the **four** points 
$$
(t_1, y_1), (t_2, y_2), (t_3, y_3), (t_4, y_4).
$$
i.e. that **minimizes** the sum 
$$
\sum_{i = 1}^4 (U(t_i) - y_i)^2.
$$
We are asked to do this using the QR factorization of the matrix $B$ found in the previous part.

If we are given a $m\times n$ inconsistent system, in matrix form as
$$
Ax = b
$$
then we can solve for the least square solution using the $QR$ factorization by setting

\begin{align*}
\hat{R} = upper\; n\times n\; submatrix\; of\; R \\ 
\hat{d} = upper\; n\; entries\; of\; d = Q^T b
\end{align*}
And so, then we need to solve the equation given as 
$$
\hat{R} \bar{x} = \hat{d} = Q^T b
$$
for $\bar{x}$ which is the least square solution to the system $Ax = b$.

